internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,188 @@
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	MaintenanceModeCRDName = "maintenancemodes.ramendr.openshift.io"
+)
+
+// MaintenanceModeReconciler reconciles a ClusterVersion object
+type MaintenanceModeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	log logr.Logger
+	ctx context.Context
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MaintenanceModeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	generationChangePredicate := predicate.GenerationChangedPredicate{}
+	storageClientStatusPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+			oldObj := e.ObjectOld.(*v1alpha1.StorageClient)
+			newObj := e.ObjectNew.(*v1alpha1.StorageClient)
+			return !reflect.DeepEqual(oldObj.Status.InMaintenanceMode, newObj.Status.InMaintenanceMode)
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).Named("Maintenance Mode").
+		Watches(
+			&ramenv1alpha1.MaintenanceMode{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate),
+		).
+		Watches(
+			&v1alpha1.StorageClaim{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate),
+		).
+		Watches(
+			&v1alpha1.StorageClient{},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(generationChangePredicate, storageClientStatusPredicate),
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes,verbs=get;list;update;create;watch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclients;storageclaims,verbs=get;list;watch
+
+func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "MaintenanceMode", req)
+	r.log.Info("Reconciling MaintenanceMode")
+
+	mapClientNameToObj := map[string]*v1alpha1.StorageClient{}
+
+	maintenanceModes := &ramenv1alpha1.MaintenanceModeList{}
+	if err := r.list(maintenanceModes); err != nil {
+		r.log.Error(err, "failed to list the Maintenance Modes")
+		return reconcile.Result{}, err
+	}
+
+	for i := range maintenanceModes.Items {
+		mm := &maintenanceModes.Items[i]
+		sc := &v1alpha1.StorageClaim{}
+		// MMode's TargetID is replicationID, which in our case is storageClaim name
+		sc.Name = mm.Spec.TargetID
+		err := r.get(sc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Use nested scopes for errors where possible
```suggestion
		if err := r.get(sc); err != nil {
			return ctrl.Result{}, err
		}
```